src/lib/homestore_backend/hs_blob_manager.cpp

```diff
@@ -281,8 +281,84 @@ homestore::blk_alloc_hints HSHomeObject::blob_put_get_blk_alloc_hints(sisl::blob
+    }
+
+    RELEASE_ASSERT(repl_dev != nullptr, "Repl dev instance null");
+
+    auto req = repl_result_ctx< BlobManager::Result< BlobInfo > >::make(sizeof(blob_id_t), io_align);
+
+    req->header_.msg_type = ReplicationMessageType::DEL_BLOB_MSG;
+    req->header_.payload_size = 0;
+    req->header_.payload_crc = 0;
+    req->header_.shard_id = shard.id;
+    req->header_.pg_id = pg_id;
+    req->header_.seal();
+    sisl::blob header;
+    header.bytes = r_cast< uint8_t* >(&req->header_);
+    header.size = sizeof(req->header_);
+
+    memcpy(req->hdr_buf_.bytes, &blob_id, sizeof(blob_id_t));
+
+    repl_dev->async_alloc_write(header, req->hdr_buf_, sisl::sg_list{}, req);
+    return req->result().deferValue([](const auto& result) -> folly::Expected< folly::Unit, BlobError > {
+        if (result.hasError()) { return folly::makeUnexpected(result.error()); }
+        auto blob_info = result.value();
+        LOGTRACEMOD(blobmgr, "Delete blob success,  shard_id {} , blob_id {}", blob_info.shard_id, blob_info.blob_id);
+
+        return folly::Unit();
+    });
+}
+
+void HSHomeObject::on_blob_del_commit(int64_t lsn, sisl::blob const& header, sisl::blob const& key,
+                                      cintrusive< homestore::repl_req_ctx >& hs_ctx) {
+    repl_result_ctx< BlobManager::Result< BlobInfo > >* ctx{nullptr};
+    if (hs_ctx != nullptr) {
+        ctx = boost::static_pointer_cast< repl_result_ctx< BlobManager::Result< BlobInfo > > >(hs_ctx).get();
+    }
+
+    auto msg_header = r_cast< ReplicationMessageHeader* >(header.bytes);
+    if (msg_header->corrupted()) {
+        LOGERROR("replication message header is corrupted with crc error, lsn:{}", lsn);
+        if (ctx) { ctx->promise_.setValue(folly::makeUnexpected(BlobError::CHECKSUM_MISMATCH)); }
+        return;
+    }
+
+    shared< BlobIndexTable > index_table;
+    shared< homestore::ReplDev > repl_dev;
+    {
+        std::shared_lock lock_guard(_pg_lock);
+        auto iter = _pg_map.find(msg_header->pg_id);
+        RELEASE_ASSERT(iter != _pg_map.end(), "PG not found");
+        index_table = static_cast< HS_PG* >(iter->second.get())->index_table_;
+        repl_dev = static_cast< HS_PG* >(iter->second.get())->repl_dev_;
+        RELEASE_ASSERT(index_table != nullptr, "Index table not intialized");
+        RELEASE_ASSERT(repl_dev != nullptr, "Repl dev instance null");
+    }
+
+    BlobInfo blob_info;
+    blob_info.shard_id = msg_header->shard_id;
+    blob_info.blob_id = *r_cast< blob_id_t* >(key.bytes);
+
+    // set a tombstone for the blob in index table.
+    // it is the GC thread's responsibility to delete the tombstone from index table and reclaim the space.
+    auto r = move_to_tombstone(index_table, blob_info);
+    if (!r) {
+        LOGW("TombStone blob not found in index id {} shard {}", blob_info.blob_id, blob_info.shard_id);
+        return;
+    }
+
+    auto& tombStone_multiBlks = r.value();
+    repl_dev->async_free_blks(lsn, tombStone_multiBlks);
```

Comment:
as an integration test, can we also verify in the test case that the space is actually freed, e.g. the m_total_freeable is increasing? Just to make sure the call made all the way to underlying chunk and its allocator

Reply:
ack

Reply:
We should avoid testing code from other components, those tests should live there yeah, especially if it results in adding a public API *just* to test HS internal logic.

Reply:
I agree on the point to avoid adding public API for just testing other component. 
I don't agree on this testing other components. The various underlying component already have their UTs to cover functional. When it is being consumed by external components, we need some level of integration test to verify the the functionality (not in lower component, but in this component this feature is being added) is actually working, right? 

We also need integration test as a full HO functionalities that HO is providing, we can't just say HO supports this, that but only before HomeStore.

Reply:
Simply, if the bug for some issue is found in repo A (or some code is not being tested) that test should live in repo A; not in dependent repo B. I'm not aware of the scenario you're trying to ensure here, it sounds like a validation for the test, but it should not be added in order to validate HomeStore is working correctly, merely that HomeObject's assumptions about the HomeStore API is valid.
In this case the value is not made available in the API (cause it's not normally useful apparently), so we don't need to add a validation here as no API invariants are called into question. If the API did already exist, i agree it's an additional validation needed since the API is what we're testing.

Reply:
Another way to put this is to treat HomeObject is only knowing that HomeStore is an API. Imagine someone implemented a version that had VChunks that did not release storage immediately upon delete (for whatever reason), now this validation is incorrect as the decreasing of this value is not part of the API contract itself.

Reply:
You misunderstood my point. My comment started with "as an integration test", so it is not to test underlying component, which already have their UT to cover its correctness. It is to test HomeObject's deletion blob api, and make sure as a public API, it does what it claims.

Reply:
@yamingk 

>1 delete it from the index. 

this has been tested in the UT of this PR

>2  free the space

this is not what we can guarantee. actually,  what we can only guarantee is that we definitely call `async_free_blks` with the correct parameter. if we call `async_free_blks` with the correct parameter but the space is not freeed as expected, it is the duty of the author of `async_free_blks` to find out the bug, right?

if we go deeper, for any syscall(e.g.  write) , after we call fd#write and sync, do we need to verify the block has been writen to disk and the block content is what we write? for every fd#write, do we need this test in our code?

i believe it is a promise from the API provider that if we call the API with correct parameter, we will get the expected result. the API should be fully tested before it is released, it is not the concern or duty of upper layer which uses it.

Reply:
Technically I believe @JacksonYao287 you only need to ensure by UT(with mock) that `async_free_blks` was called with proper parameters.

Practically as https://github.com/eBay/HomeStore/pull/216/files already expose the `freeable_blks` to vchunk,  having an integration test is fine.

Reply:
If we can have mock inside vchunk it would be good enough I think for this PR.